### PR TITLE
Fast Reduce

### PR DESCRIPTION
**Pull request by aszepieniec**
This PR
 - Use clever math to accelerate modular reduction of one polynomial by another, especially when the modulus has much lower degree.
 - Test the introduced methods.
 - Integrate the fast modular reduction into the batch evaluation dispatcher pipeline, resulting in faster batch evaluation.
 - Separate the parallel from sequential code paths.

### CODE REVIEW

File: twenty-first/src/math/polynomial.rs [hunk 1]
```diff
@@ -765,11 +885,378 @@ where
...
+    /// g(X) * f(X) = 1 mod X^n , where n is the precision.
+    ///
+    /// In formal terms, g(X) is the approximate multiplicative inverse in
+    /// the formal power series ring, where elements obey the same
+    /// algebraic rules as polynomials do but can have an infinite number of
+    /// coefficients. To represent these elements on a computer, one has to
+    /// truncate the coefficient vectors somewhere. The resulting truncation
+    /// error is considered "small" when it lives on large powers of X. This
+    /// function works by applying Newton's method in this ring.
+    ///
```
Comment (jan-ferdinand) [inline twenty-first/src/math/polynomial.rs:907]:
```suggestion
    ///
    /// # Example
    ///
    /// ```
    /// # use num_traits::One;
    /// # use twenty_first::prelude::*;
    /// let precision = 8;
    /// let f = Polynomial::new(bfe_vec![42; precision]);
    /// let g = f.formal_power_series_inverse_newton(precision);
    /// let x_to_the_n = Polynomial::one().shift_coefficients(precision);
    /// let (_quotient, remainder) = g.multiply(&f).divide(&x_to_the_n);
    /// assert!(remainder.is_one());
    /// ```
```

File: twenty-first/src/math/polynomial.rs [hunk 2]
```diff
@@ -2583,4 +3084,396 @@ mod test_polynomials {
...
+
+    #[proptest]
+    fn reduce_by_ntt_friendly_modulus_and_reduce_long_division_agree(
+        #[strategy(1usize..10)] m: usize,
+        #[strategy(vec(arb(), #m))] b_coefficients: Vec<BFieldElement>,
+        #[strategy(1usize..100)] _deg_a: usize,
+        #[strategy(vec(arb(), #_deg_a + 1))] _a_coefficients: Vec<BFieldElement>,
+        #[strategy(Just(Polynomial::new(#_a_coefficients)))] a: Polynomial<BFieldElement>,
+    ) {
+        if !b_coefficients.iter().all(|c| c.is_zero()) {
```
Comment (jan-ferdinand) [inline twenty-first/src/math/polynomial.rs:3373]: This is confusing me. Is this checking whether polynomial `b` would be zero? If so, how about
```suggestion
		let b = Polynomial::new(b_coefficients);
		if b.is_zero() {            
			return Err(TestCaseError::Reject("some reason".into()));
		}
```
or an equivalent `#[filter(…)]`?

File: twenty-first/src/math/polynomial.rs [hunk 3]
```diff
@@ -2583,4 +3084,396 @@ mod test_polynomials {
...
+    ) {
+        let standard_remainder = a.reduce_long_division(&b);
+        let preprocessed_remainder = a.fast_reduce(&b);
+        prop_assert_eq!(standard_remainder, preprocessed_remainder);
+    }
+
+    #[test]
+    fn fast_reduce_agrees_with_reduce_long_division_on_huge_polynomials() {
+        // This test could be made into a proptest but to make test time
+        // manageable we would have to set cases=1. And then what's the point?
```
[the remark this answers is not in the document]
Comment (aszepieniec) [reply twenty-first/src/math/polynomial.rs:3445]: Dropped for time.

File: twenty-first/src/math/polynomial.rs [hunk 4]
```diff
@@ -656,9 +749,43 @@ where
...
     pub fn batch_evaluate(&self, domain: &[FF]) -> Vec<FF> {
-        // According to `cargo bench --bench evaluation` on mjolnir, parallel evaluation is always
-        // faster than fast evaluation.
+        if self.degree() >= Self::REDUCE_BEFORE_EVALUATE_THRESHOLD_RATIO * (domain.len() as isize) {
+            self.reduce_then_batch_evaluate(domain)
+        } else if domain.len() >= Self::DIVIDE_AND_CONQUER_EVALUATE_CUTOFF_THRESHOLD {
+            self.divide_and_conquer_batch_evaluate(domain)
+        } else {
+            self.iterative_batch_evaluate(domain)
+        }
```
[the remark this answers is not in the document]
Comment (aszepieniec) [reply twenty-first/src/math/polynomial.rs:764]: This method is substantially revised in the upcoming PR, `asz/fast-extrapolate`.

**Pull request merged**